design-proposals/klip-25-removal-of-with-key-syntax.md

```diff
@@ -0,0 +1,163 @@
+
+-- is less expressive than --
+
+CREATE TABLE USERS_BY_REGION AS
+   SELECT REGION, COUNT(*) FROM USERS GROUP BY REGION;
+```
+
+However, [a feature removing the restriction that key columns must be called `ROWKEY` is about to be
+enabled](https://github.com/confluentinc/ksql/pull/5093). Once it is, the aliasing that `WITH(KEY)`
+provides will be redundant, as the same can be achieved much more intuitively, and without the need
+to duplicate the key data into a value field, using the new syntax. For example:
+
+```sql
+-- old skool syntax:
+CREATE TABLE USERS (ROWKEY BIGINT PRIMARY KEY, NAME STRING, ID BIGINT) WITH ('KEY'='ID', ...);
+
+-- new skool syntax:
+CREATE TABLE USERS (ID BIGINT PRIMARY KEY, NAME STRING) WITH (...);
+```
+
+### Optimisations
+
+This aliasing allows ksqlDB to optimize certain queries by avoiding unnecessary repartition steps,
+should the 'key field' be used in a `PARTITION BY`, `GROUP BY` or `JOIN ON` clause. For example:
+
+```sql
+-- given:
+CREATE TABLE INPUT (ID INT, NAME STRING) WITH ('KEY'='ID', 'KAFKA_TOPIC'='input', ...);
+
+-- no repartition necessary as `ID` is equivalent to `ROWKEY`
+SELECT ID, COUNT() FROM INPUT GROUP BY ID;
+```
+
+KsqlDB can avoid an unnecessary repartition step in the above statement as `ID` is equivalent to
+`ROWKEY`, and grouping by `ROWKEY` is already correctly partitioned.
+
+It's not just streams and tables created by `CREATE STREAM` and `CREATE TABLE` statements that can
+track these `key-fields`. KsqDB also tries to keep track of the key field in derived sources
+too, for example:
+
+```sql
+CREATE TABLE FOO AS ID, COUNT() FROM BAR GROUP BY ID;
+-- Foo has key-field 'ID'
+```
+
+In the above, KsqlDB will detect that `FOO``s `ID` column is equivalent to `ROWKEY`. However, it can
+only handle a single field. If there are multiple, only the first is tracked, for example:
+
+```sql
+CREATE TABLE FOO AS ID AS ID1, ID AS ID2, COUNT() FROM BAR GROUP BY ID;
+-- Foo has key-field 'ID1'
+-- `ID2` is not a key field, even though its a duplicate of `ID1`
+```
+
+Unfortunately, the implementation of the handling of these 'key fields' is very buggy, full of edge
+cases and incomplete. The bugs and edgecases can mean that a repartition is done when not needed or,
+worse, not done when needed. The incomplete implementation can result in confusing behaviour for
+users.
+
+With the introduction of the 'any-key-name' feature, this optimization also becomes redundant, as
+users can directly name their key columns as they need.
+
+## What is in scope
+
+- Removal of `WITH(KEY)` from ksqlDB syntax
+- Removal of the code with ksqlDB that tracks the key-field.
+
+## What is not in scope
+
+- Future `WITH(KEYS)` style syntax that may be used to control which columns are persisted in the
+  key of the Kafka record vs the value.  This is TBD and would be covered in a separate KLIP.
+
+## Value/Return
+
+The feature is not intuitive, it's the source of many a confused user and engineer and is often
```

Comment:
Nit: `The current WITH(key) feature`

(Sounds like "the new feature is not intuitive" atm.)